### PR TITLE
feat(tui): add execution history navigation and snapshots

### LINKED ISSUE #217: TUI: Execution History for Watch Mode
## Summary
Add per-stage execution history to the TUI watch mode. Each stage re-run creates an in-memory history entry. Users navigate between executions with j/k arrows while viewing a single unified view.

## Key Decisions
- **Storage**: In-memory only (lost when TUI closes)
- **Navigation**: j/k moves between past executions, arrows indicate more entries
- **Run ID**: Use backend's `run_id` from `TuiStatusMessage` (add to message type)
- **UI Pattern**: Single unified view with execution indicator, no separate "list mode"

---

### Design Principle: Unified View with Navigation

There's only one view state. The current execution (live or historical) is always displayed. Arrow indicators and execution numbers show there are other entries available.

### Normal View (Live, No History Yet)

```
┌─ Detail [process_data] ─────────────────────────────────────────┐
│ ┌ Logs │ Input │ Output ┐                      [1/1] ● LIVE     │
│ │                                                               │
│ │ [14:23:15] Processing data...                                 │
│ │ [14:23:16] Loaded 1000 rows                                   │
│ │ [14:23:17] Complete                                           │
│                                                                 │
└─────────────────────────────────────────────────────────────────┘
```
- `[1/1]` - execution 1 of 1 (only one so far)
- `● LIVE` - green indicator, logs streaming

### View with History Available (Viewing Live)

```
┌─ Detail [process_data] ─────────────────────────────────────────┐
│ ┌ Logs │ Input │ Output ┐                   ← [3/3] ● LIVE      │
│ │                                                               │
│ │ [14:23:15] Processing data...                                 │
│ │ [14:23:16] Loaded 1000 rows                                   │
│ │ [14:23:17] Complete                                           │
│                                                                 │
│                                                  j: prev exec   │
└─────────────────────────────────────────────────────────────────┘
```
- `←` left arrow indicates older executions available (press j)
- `[3/3]` - viewing execution 3 of 3 (most recent/live)
- Footer hint shows `j: prev exec`

### Viewing Historical Execution

Press `j` to go back:

```
┌─ Detail [process_data] ─────────────────────────────────────────┐
│ ┌ Logs │ Input │ Output ┐        ← [2/3] ◷ 14:20:01 (2.1s) → │
│ │                                                               │
│ │ [14:20:01] Processing data...                                 │
│ │ [14:20:02] Loaded 950 rows                                    │
│ │ [14:20:03] Complete                                           │
│ │                                                               │
│                                      j: prev  k: next  q: live  │
└─────────────────────────────────────────────────────────────────┘
```
- `←` left arrow: older entries exist (press j)
- `→` right arrow: newer entries exist (press k)
- `[2/3]` - viewing execution 2 of 3
- `◷ 14:20:01 (2.1s)` - timestamp and duration (amber, not live)
- Footer shows navigation hints

### Viewing Oldest Execution

```
┌─ Detail [process_data] ─────────────────────────────────────────┐
│ ┌ Logs │ Input │ Output ┐           [1/3] ◷ 14:15:33 (0.0s) ✗ →│
│ │                                                               │
│ │ [14:15:33] Processing data...                                 │
│ │ [14:15:33] ERROR: FileNotFoundError: data.csv                 │
│ │                                                               │
│ │                                                               │
│                                             k: next  q: live    │
└─────────────────────────────────────────────────────────────────┘
```
- No `←` arrow (this is the oldest)
- `→` arrow indicates newer entries (press k)
- `✗` status indicator for failed execution

### Full History List (Shift+H)

Press `Shift+H` to see full list overlay:

```
┌─ Detail [process_data] ─────────────────────────────────────────┐
│ ┌ Logs │ Input │ Output ┐                   ← [2/3] ◷ 14:20:01 →│
│ ├───────────────────────────────────────────────────────────────┤
│ │  #  │ Time     │ Duration │ Status  │ Reason                  │
│ │─────┼──────────┼──────────┼─────────┼─────────────────────────│
│ │   3 │ 14:23:15 │     2.3s │ ✓ SUCCESS│ code changed           │
│ │ ▸ 2 │ 14:20:01 │     2.1s │ ✓ SUCCESS│ input changed          │
│ │   1 │ 14:15:33 │     0.0s │ ✗ FAILED │ FileNotFoundError      │
│ ├───────────────────────────────────────────────────────────────┤
│ │ j/k: navigate  Enter: view  Esc: close                        │
└─────────────────────────────────────────────────────────────────┘
```
- Full list shows all executions
- `▸` marker on currently viewed entry
- j/k navigates, Enter jumps to entry, Esc closes list

### Navigation Bindings

| Key | Effect |
|-----|--------|
| `j` | View previous (older) execution |
| `k` | View next (newer) execution |
| `q` | Jump to live view (most recent) |
| `Shift+H` | Open/close full history list |
| `h/l, Tab` | Switch tabs (Logs/Input/Output) |

---

## Indicator Summary

| Element | Meaning |
|---------|---------|
| `● LIVE` | Viewing live execution (green) |
| `◷ HH:MM:SS (Xs)` | Viewing historical execution (amber) |
| `←` | Older executions available (press j) |
| `→` | Newer executions available (press k) |
| `[N/M]` | Viewing execution N of M total |
| `✓` | Success |
| `✗` | Failed |
| `⊘` | Skipped |

---

### ExecutionHistoryEntry (new dataclass in run.py)

```python
@dataclasses.dataclass
class ExecutionHistoryEntry:
    run_id: str                   # From TuiStatusMessage, format: YYYYMMDD_HHMMSS_<uuid8>
    stage_name: str
    timestamp: float              # time.time() at execution start
    duration: float | None        # seconds, None if still running
    status: StageStatus
    reason: str
    logs: list[tuple[str, bool, float]]  # (line, is_stderr, ts) - no limit, full log capture
    input_snapshot: StageExplanation | None
    output_snapshot: list[OutputChange] | None
```

### TuiStatusMessage modification (types.py)

Add `run_id` field to `TuiStatusMessage`:
```python
class TuiStatusMessage(TypedDict):
    type: Literal[TuiMessageType.STATUS]
    stage: str
    index: int
    total: int
    status: StageStatus
    reason: str
    elapsed: float | None
    run_id: str  # NEW: Pipeline run ID from executor
```

### StageInfo modifications (run.py:84-96)

```python
history: collections.deque[ExecutionHistoryEntry] = dataclasses.field(
    default_factory=lambda: collections.deque(maxlen=50)
)
```

### WatchTuiApp state

```python
_viewing_history_id: str | None = None  # None = live view, else = run_id of historical entry
_history_list_open: bool = False        # True when Shift+H overlay is shown
```

---

### 0. Add run_id to TuiStatusMessage (backend change)
- **File**: `src/pivot/types.py`
- Add `run_id: str` field to `TuiStatusMessage`
- **File**: `src/pivot/executor/core.py`
- Pass `run_id` when creating `TuiStatusMessage` in `_send_tui_status()`

### 1. Data structures and types
- **File**: `src/pivot/tui/run.py`
- Add `ExecutionHistoryEntry` dataclass with `run_id` from status message
- Extend `StageInfo` with `history` deque (maxlen=50)
- Add state to `WatchTuiApp`: `_viewing_history_id`, `_history_list_open`

### 2. History capture on status changes
- **File**: `src/pivot/tui/run.py`
- Add methods to `WatchTuiApp`:
  - `_create_history_entry(stage_name, run_id)` - on IN_PROGRESS, capture input snapshot, start accumulating logs
  - `_finalize_history_entry(stage_name, status, reason, elapsed)` - on terminal status, capture output snapshot
- **Error handling**: Wrap snapshot capture in try/except; set snapshot to None on failure
- **Log capture**: Accumulate logs separately during execution (not from the bounded deque)
  - The live `StageInfo.logs` deque keeps `maxlen=1000` for performance
  - History entries accumulate ALL logs in a separate list during execution
  - On finalization, move accumulated logs to the `ExecutionHistoryEntry`

### 3. Header indicator rendering
- **File**: `src/pivot/tui/run.py`
- Modify `TabbedDetailPanel` header to show:
  - Arrow indicators (`←` / `→`) when other entries exist
  - Execution count `[N/M]`
  - Mode: `● LIVE` (green) or `◷ HH:MM:SS (Xs) ✓/✗` (amber)
- Add footer hint: `j: prev exec` when history exists

### 4. Navigation integration
- **File**: `src/pivot/tui/run.py`
- Modify `action_nav_up/down` (j/k) to navigate history when in detail panel with history
- Add `action_goto_live()` bound to `q` - clears `_viewing_history_id`
- When viewing historical entry, tabs show that entry's snapshot data

### 5. History list overlay (Shift+H)
- **File**: `src/pivot/tui/run.py`
- Add `HistoryListOverlay` widget that renders over tab content
- Add `H` binding: `action_toggle_history_list()`
- j/k navigates list, Enter jumps to entry, Esc closes

### 6. Panel snapshot rendering
- **File**: `src/pivot/tui/diff_panels.py`
- Add `set_from_snapshot()` methods:
  - `InputDiffPanel.set_from_snapshot(stage_name, snapshot: StageExplanation)`
  - `OutputDiffPanel.set_from_snapshot(stage_name, snapshot: list[OutputChange])`
- **File**: `src/pivot/tui/run.py`
- Add `StageLogPanel.set_from_history(logs: list[tuple[str, bool, float]])`

### 7. Tests
- **File**: `tests/tui/test_history.py` (new)
- Unit tests:
  - ExecutionHistoryEntry creation with run_id
  - StageInfo.history bounded at 50 entries
  - History ID survives deque rotation
  - Snapshot capture failure handled gracefully
- Integration tests:
  - History entry created on stage completion
  - j/k navigates between executions
  - Arrow indicators appear correctly
  - Tab switches show historical Logs/Input/Output
  - Shift+H opens/closes history list
  - q returns to live view

---

## Critical Implementation Notes

1. **Use run_id for identification** - Use `run_id` from `TuiStatusMessage` for `_viewing_history_id`
2. **Error handling is mandatory** - Wrap `explain.get_stage_explanation()` in try/except; log failures, never crash
3. **Full log capture for history** - Accumulate ALL logs during execution for history entries; live view keeps `maxlen=1000` deque
4. **Arrow logic**:
   - Show `←` when viewing entry N where N > 1
   - Show `→` when viewing entry N where N < total
   - When live: show `←` if history has entries, no `→`
5. **Guard against edge cases**:
   - No stage selected: disable j/k history navigation
   - Empty history: show `[1/1] ● LIVE` with no arrows
   - Stage removed during history view: exit to live

---

## Files to Modify

| File | Changes |
|------|---------|
| `src/pivot/types.py` | Add `run_id: str` to `TuiStatusMessage` |
| `src/pivot/executor/core.py` | Pass `run_id` when creating `TuiStatusMessage` |
| `src/pivot/tui/run.py` | ExecutionHistoryEntry, StageInfo.history, header indicators, j/k history nav, HistoryListOverlay |
| `src/pivot/tui/diff_panels.py` | `set_from_snapshot()` methods on InputDiffPanel, OutputDiffPanel |
| `tests/tui/test_history.py` | New test file for history feature |

---

## Verification

1. **Manual testing**:
   - Run `pivot watch` on a project with stages
   - Trigger stage re-runs (modify input files)
   - Verify `[N/M]` counter updates and arrows appear
   - Press j to go back, k to go forward, q to return to live
   - Press Shift+H to see full history list
   - Switch tabs while viewing history to see historical Input/Output

2. **Automated tests**:
   - `uv run pytest tests/tui/test_history.py -v`
   - `uv run ruff format . && uv run ruff check . && uv run basedpyright .`

### PR DESCRIPTION
**Pull request by sjawhar**
## Summary
- Add history navigation keybindings (`[` / `]` / `G`) to browse older/newer executions
- Add history list modal (`H`) with j/k navigation for direct jump to any execution
- Capture input/output snapshots at execution time for historical views in Input/Output tabs
- Fix panel ID bug in `_update_history_view` that prevented snapshot display

Closes #217

## Changes
| File | Description |
|------|-------------|
| `src/pivot/tui/run.py` | Keybindings, `HistoryListScreen` modal, snapshot capture in `_create_history_entry`/`_finalize_history_entry` |
| `src/pivot/tui/diff_panels.py` | `set_from_snapshot()` methods for `InputDiffPanel` and `OutputDiffPanel` |
| `src/pivot/executor/core.py` | Add `run_id` to status messages |
| `tests/tui/test_history.py` | New tests for history data structures |

## Test plan
- [x] All 2432 tests pass
- [x] Type checker passes with 0 errors
- [x] Manual testing: `pivot watch`, trigger re-runs, navigate history with `[`/`]`/`G`/`H`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by sjawhar** [comment]
## Code Analysis Findings

Ran `/analyze` on the StageLifecycle refactoring and TUI history changes. Here are the findings for follow-up:

### High Priority

| Issue | File | Description |
|-------|------|-------------|
| **TUI queue blocking** | `core.py:146-158` | `tui_queue.put()` can block indefinitely if TUI crashes/hangs. Consider `put_nowait()` with exception handling. |

### Medium Priority

| Issue | File | Description |
|-------|------|-------------|
| **Memory leak on stage removal** | `run.py:1602-1606` | Pending history entries not cleaned up when stage is removed mid-execution. Pop `_pending_history[stage_name]` at start of `_finalize_history_entry`. |
| **Thin wrapper function** | `diff_panels.py:117-119` | `_get_cache_dir()` now just delegates to `project.get_cache_dir()`. Could inline directly. |
| **Dead fallback branches** | `core.py:872-879, 918-925` | `lifecycle` parameter is always provided in practice; `else` branches may never execute. Consider making required or adding tests. |

### Low Priority (Deferred)

| Issue | File | Description |
|-------|------|-------------|
| Hardcoded `completed_count=0` in `mark_skipped_upstream` | `core.py:196` | Cosmetic - shows wrong index for skipped stages in console |
| History index on stage removal | `run.py:1533-1536` | `_selected_stage_name` not reset when stage removed |
| Silent exception swallowing | `run.py:1580` | Exceptions logged without details |
| Multi-cell character alignment | `diff_panels.py:72-79` | Documented limitation with CJK/emoji |

### What's Working Well
- ✅ StageLifecycle abstraction eliminates notification duplication
- ✅ Bounded deques prevent memory growth (`maxlen=500` logs, `maxlen=50` history)  
- ✅ Run ID tracking clears stale pending entries on new run
- ✅ All tests passing (2448 passed), 0 type errors

File: src/pivot/tui/diff_panels.py [hunk 1]
```diff
@@ -121,7 +131,7 @@ def _get_relative_path(abs_path: str) -> str:
     return abs_path
 
 
-def _compute_output_changes(
+def compute_output_changes(
```
Comment (sjawhar) [inline src/pivot/tui/diff_panels.py]: This function was renamed from `_compute_output_changes` to `compute_output_changes` (dropping the leading underscore), making it public API. Is this intentional? If so, should it be added to `__all__`?

File: src/pivot/tui/run.py [hunk 2]
```diff
@@ -24,14 +24,18 @@
 import textual.timer
 import textual.widgets
 
-from pivot import project
+from pivot import explain, parameters, project
 from pivot.executor import ExecutionSummary
 from pivot.executor import commit as commit_mod
+from pivot.registry import REGISTRY
```
Comment (sjawhar) [inline src/pivot/tui/run.py]: Import style violation per CLAUDE.md: "Import modules, not functions". This should be `from pivot import registry` then use `registry.REGISTRY`. The same applies to `from pivot.tui import diff_panels` on line 32 - good, that one follows the convention correctly.

File: src/pivot/tui/run.py [hunk 3]
```diff
@@ -81,6 +85,34 @@ def _format_elapsed(elapsed: float | None) -> str:
...
+    logs: list[tuple[str, bool, float]]
+    input_snapshot: StageExplanation | None
+    output_snapshot: list[OutputChange] | None
+
+
+@dataclasses.dataclass
+class _PendingHistoryState:
+    """Temporary state for a stage execution in progress, before finalization."""
+
+    run_id: str
```
Comment (sjawhar) [inline src/pivot/tui/run.py]: Good: bounded deques (`maxlen=500` for logs, `maxlen=50` for history) prevent memory growth in watch mode with verbose stages. Nice attention to resource management.

File: src/pivot/tui/run.py [hunk 4]
```diff
@@ -280,13 +324,18 @@ class TabbedDetailPanel(textual.containers.Vertical):
     """Tabbed panel showing stage details with Logs, Input, Output tabs."""
 
     _stage: StageInfo | None
+    _history_index: int | None  # None = live view, else index into history deque
```
Comment (sjawhar) [inline src/pivot/tui/run.py]: Naming inconsistency: `_history_index` here vs `_viewing_history_index` in `WatchTuiApp` (line 1393). Consider using consistent naming across the codebase for the same concept.

File: src/pivot/types.py [hunk 5]
```diff
@@ -460,6 +460,7 @@ class TuiStatusMessage(TypedDict):
     status: StageStatus
     reason: str
     elapsed: float | None
+    run_id: str
```
Comment (sjawhar) [inline src/pivot/types.py]: `run_id: str` is now a required field on TuiStatusMessage. Please verify all code paths that create TuiStatusMessage include `run_id`, especially the timeout handling path in `_execute_greedy` which uses `_mark_stage_failed`.

File: tests/execution/test_lifecycle.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
+import multiprocessing as mp
```
Comment (sjawhar) [inline tests/execution/test_lifecycle.py]: Good test coverage for the new StageLifecycle class. Tests cover all state transitions (started, completed, failed, skipped_upstream) and verify both state updates and TUI message sending.

File: tests/tui/test_history.py [hunk 7]
```diff
@@ -0,0 +1,389 @@
+"""Tests for TUI execution history feature."""
```
Comment (sjawhar) [inline tests/tui/test_history.py]: Good edge case coverage: tests for bounded deque limits (500 for pending logs, 50 for history entries), separate instance verification, and the critical fix for skipped stages without pending state getting history entries.
